walnut/validate/validate.go

```diff
@@ -0,0 +1,100 @@
+/*
+ * === This file is part of ALICE O² ===
+ *
+ * Copyright 2020 CERN and copyright holders of ALICE O².
+ * Author: Ayaan Zaidi <azaidi@cern.ch>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * In applying this license CERN does not waive the privileges and
+ * immunities granted to it by virtue of its status as an
+ * Intergovernmental Organization or submit itself to any jurisdiction.
+ */
+
+package validate
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/AliceO2Group/Control/walnut/schemata"
+
+	"github.com/xeipuuv/gojsonschema"
+	"gopkg.in/yaml.v2"
+)
+
+// CheckSchema accepts a filename and format then validate against the schema specified (either workflow or task)
+func CheckSchema(filename string, format string) {
+
+	rawYAML, err := ioutil.ReadFile(filename) // import YAML file
+	if err != nil {
+		fmt.Printf("ReadFile failed: %v", err)
+		os.Exit(1)
+	}
+
+	var dataFromYAML interface{}
+	if err := yaml.Unmarshal([]byte(rawYAML), &dataFromYAML); err != nil {
+		fmt.Printf("Unmarshaling YAML failed: %v", err)
```

Comment:
Instead of `fmt.Printf` here, I advise returning the same message in a wrapped error, i.e. `return fmt.Errorf("Unmarshaling YAML failed: %w", err)`. The caller should then check the error, print as needed and `os.Exit`.

See https://blog.golang.org/go1.13-errors for more information on error wrapping.

A function like `CheckSchema` should never `os.Exit(1)`, in command line utilities one has to be careful because it's very easy to lose track. It should return `error` to the caller, which should check whether the `error` is `nil` and act accordingly.

You generally want your `os.Exit` calls to be few and in places where you control exactly what and how you print out the error and the exit code.

Reply:
Understood.